### ISSUE TITLE
StaticTypeRulesDiagnosticAnalyzer implements CA1053 incorrectly

### ISSUE THREAD

**Issue by jmarolf** [dotnet/roslyn#2033]
[CA1053](https://msdn.microsoft.com/en-us/library/ms182169.aspx) should trigger if there are _any_ public constructors on a type that only has static member.

So the following code should trigger CA1053 because the compiler generates a public constructor:
```C#
using System;

namespace DesignLibrary
{
   public class NoInstancesNeeded
   {
      // Violates rule: StaticHolderTypesShouldNotHaveConstructors. 
      // Uncomment the following line to correct the violation. 
      // private NoInstancesNeeded() {} 

      public static void Method1() {}
      public static void Method2() {}
   }
}
```

CA1053 should also trigger here:
```C#
using System;

namespace DesignLibrary
{
    public class NoInstancesNeeded
    {
        public NoInstancesNeeded() {} 
        public static void Method1() { }
        public static void Method2() { }
    }
}
```

Neither of these scenarios produce an error.  The only case in which [StaticTypeRulesDiagnosticAnalyzer](http://source.roslyn.io/#Microsoft.CodeAnalysis.FxCopAnalyzers/Design/StaticTypeRulesDiagnosticAnalyzer.cs) will give CA1053 is if the type has a static constructor and a public constructor.
```C#
using System;

namespace DesignLibrary
{
    public class NoInstancesNeeded
    {
        static NoInstancesNeeded() {} 
        public NoInstancesNeeded(int i) { }
        public static void Method1() { }
        public static void Method2() { }
    }
}
```

**Comment by lgolding** [comment]
Before I work on this, let's discuss it a little more. The wording of [CA1053](https://msdn.microsoft.com/en-us/library/ms182169.aspx) suggests to me that they were worried specifically about the case where you didn't author a constructor, and so the compiler gave you a default constructor, and so you unintentionally ended up with a class you could instantiate. Both the **Cause** and the **How to Fix Violations** section of that article specify that CA1053 is targeted at default constructors.

If you explicitly author

````csharp
class Foo
{
    Foo(int i) {}
    static void Bar() {}
}
````

... then maybe you had something in mind, but you just haven't finished the code yet.

Please let me know if you've already been through this discussion and decided that this really is a change you want.

**Comment by gafter** [comment]
The documentation for this diagnostic clearly reflects that the suggested change should be made. I agree that it appears targeted for default constructors only.

I recommend you try the previous implementation to determine its behavior in the scenarios you are asking about. We should aim to reproduce the prior behavior except where we believe there are bugs.

/cc @gafter @jmarolf @srivatsn

**Comment by lgolding** [comment]
After discussion with @gafter and @jmarolf, we agree that we should match FxCop CA1053 behavior exactly, which is:

- Warn if only the compiler-generated public default constructor is present.
- Warn if the user authored a public default constructor.
- Warn if the user authored a protected default constructor.
- Do not warn if the user authored a non-default constructor.

/cc @srivatsn
